feat(abs): add operator logic

vdice opened the pull request:
(Re-issue of https://github.com/coreos/etcd-operator/pull/1369)

TODO:
 - [x] Revise approach on passing storage credentials to pod (currently uses non-thread-safe `os.Setenv()`)

vdice commented:
@hongchaodeng ready to tackle the TODO item above -- wondering if you had a specific approach in mind?

vdice commented on pkg/cluster/backup_manager.go:
```diff
@@ -91,6 +92,11 @@ func (bm *backupManager) setupStorage() (s backupstorage.Storage, err error) {
 			return nil, errNoS3ConfigForBackup
 		}
 		s, err = backupstorage.NewS3Storage(c.S3Context, c.KubeCli, cl.Name, cl.Namespace, *b)
+	case spec.BackupStorageTypeABS:
```
@hongchaodeng in addition, I seem to have lost clarity on the path of storage backend configuration through the code on its way to actually creating backups.  

For instance, what exactly is being set up in this stanza and how does it compare/tie-in with the sidecar deployment setup further below?  I found myself thinking perhaps we only need the sidecar deployment setup (`k8sutil.AttachABSToPodSpec(..)`) but alas the etcd cluster resource can't be created if I remove the logic here...

hongchaodeng replied:
The logical storage components looks like this:

- backup manager: This sits on operator. It abstracts backup sidecar related logic (k8s)  for creating deployments, and backup storage specific logic (S3/ABS) for cloning data for restore and so on.
- backup sidecar: This is the backup sidecar that only takes care of saving/serving snapshots.

So basically the storage logic has been split up into two places. We can't abstract away all logic into backup sidecar.

vdice commented:
@hongchaodeng no more `os.Setenv(..)` in latest changes (now just passing data between functions).  Let me know if this looks good at your convenience...

hongchaodeng commented:
@vdice 
Can you rebase and push again?
The unit test failure was due to recent change.

fanminshi commented on pkg/backup/abs/absconfig/abs_ctx.go:
```diff
@@ -0,0 +1,21 @@
+package absconfig
+
+// ABSContext represents the controller context around a ABS backend configuration
+type ABSContext struct {
```
I don't think `ABSContext` is needed since it is never populated from anywhere; We are not parsing any `ABS` flags from etcd-operator binary.

All the ABS information should be passed in by cluster spec via `spec.BackupPolicy`.

 cc/ @hongchaodeng for confirmation.

vdice replied:
Okay, will wait for confirmation before removing...

fanminshi commented on pkg/cluster/backupstorage/abs.go:
```diff
@@ -0,0 +1,104 @@
+)
+
+type abs struct {
+	absconfig.ABSContext
```
I don't think `absconfig.ABSContext` is needed.

fanminshi commented on pkg/cluster/backupstorage/abs.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+// NewABSStorage returns a new ABS Storage implementation using the given absCtx, kubecli, cluster name, namespace and backup policy
+func NewABSStorage(absCtx absconfig.ABSContext, kubecli kubernetes.Interface, clusterName, ns string, p spec.BackupPolicy) (Storage, error) {
```
No need to have `absCtx absconfig.ABSContext`

fanminshi commented on pkg/cluster/backupstorage/abs.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	ws := &abs{
+		ABSContext:   absCtx,
```
No need to have `ABSContext:   absCtx`

fanminshi commented on pkg/cluster/cluster.go:
```diff
@@ -61,6 +62,7 @@ type Config struct {
 	PVProvisioner  string
 	ServiceAccount string
 	s3config.S3Context
+	absconfig.ABSContext
```
No need.

fanminshi commented:
@vdice have you manually test this code? and what's your manual testing plan?

vdice commented:
@fanminshi I have manually tested this code.  Boils down to 3 stages...

### Image Build/Push

First, I build the image from this branch:
```
 $ IMAGE=quay.io/vdice/etcd-operator:abs-testing hack/build/operator/build
```
### Azure setup

Next, after creating an Azure storage account, I have the account name and account key that I can provide via the k8s secret as mentioned in https://github.com/coreos/etcd-operator/blob/master/doc/design/abs_backup.md#in-detail.  

I then create the ABS container, as also mentioned in that doc.

### K8s setup/deployment

Now, on a k8s 1.7+ minikube cluster, I deploy etcd-operator via [helm](https://github.com/kubernetes/helm) installing the [chart](https://github.com/kubernetes/charts/tree/master/stable/etcd-operator) with the following values:
```
$ helm install stable/etcd-operator --name etcd-operator \
  --set image.repository="quay.io/vdice/etcd-operator" \
  --set image.tag="abs-testing"
```
I can then create the k8s secret and etcd cluster resources via manifests (as mentioned in doc as well.)  Here are some examples:
```
$ cat abs-credentials.yaml
apiVersion: v1
kind: Secret
metadata:
  name: abs-credentials
type: Opaque
stringData:
  storage-account: my-storage-account-name
  storage-key: my-storage-account-key

$ cat example-etcd-cluster.yaml
apiVersion: "etcd.database.coreos.com/v1beta2"
kind: "EtcdCluster"
metadata:
  name: etcd-cluster
spec:
  size: 1
  backup:
    backupIntervalInSecond: 10
    maxBackups: 5
    autoDelete: false
    storageType: "ABS"
    abs:
      absContainer: vdice-etcd-backups
      absSecret: abs-credentials

$ kubectl create -f abs-credentials.yaml
secret "abs-credentials" created

$ kubectl create -f example-etcd-cluster.yaml
etcdcluster "etcd-cluster" created
```

Next, we can inspect the pods once they are created and running:
```
 $ kubectl get po
NAME                                           READY     STATUS    RESTARTS   AGE
etcd-cluster-0000                              1/1       Running   0          8m
etcd-cluster-backup-sidecar-3289547686-266b4   1/1       Running   0          8m
etcd-operator-etcd-operator-1321659961-r6x9l   1/1       Running   0          8m

 $ kubectl logs -f etcd-operator-etcd-operator-1321659961-r6x9l
time="2017-09-07T23:51:16Z" level=info msg="etcd-operator Version: 0.5.1+git"
time="2017-09-07T23:51:16Z" level=info msg="Git SHA: 3d45d08"
time="2017-09-07T23:51:16Z" level=info msg="Go Version: go1.8.3"
time="2017-09-07T23:51:16Z" level=info msg="Go OS/Arch: linux/amd64"
time="2017-09-07T23:51:16Z" level=info msg="Event(v1.ObjectReference{Kind:"Endpoints", Namespace:"default", Name:"etcd-operator", UID:"70415c10-9427-11e7-a9c4-5ede2b14b2b2", APIVersion:"v1", ResourceVersion:"1159", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' etcd-operator-etcd-operator-1321659961-r6x9l became leader"
time="2017-09-07T23:51:16Z" level=info msg="finding existing clusters..." pkg=controller
time="2017-09-07T23:51:16Z" level=info msg="starts running from watch version: 1161" pkg=controller
time="2017-09-07T23:51:16Z" level=info msg="start watching at 1161" pkg=controller
time="2017-09-07T23:51:17Z" level=info msg="creating cluster with Spec:" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="{" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="    "size": 1," cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="    "baseImage": "quay.io/coreos/etcd"," cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="    "version": "3.1.8"," cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="    "backup": {" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="        "storageType": "ABS"," cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="        "abs": {" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="            "absContainer": "vdice-etcd-backups"," cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="            "absSecret": "abs-credentials"" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="        }," cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="        "backupIntervalInSecond": 10," cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="        "maxBackups": 5," cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="        "autoDelete": false" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="    }" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="}" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="backup sidecar deployment and service created" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="cluster created with seed member (etcd-cluster-0000)" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:17Z" level=info msg="start running..." cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:25Z" level=info msg="Start reconciling" cluster-name=etcd-cluster pkg=cluster
time="2017-09-07T23:51:25Z" level=info msg="Finish reconciling" cluster-name=etcd-cluster pkg=cluster
...

 $ kubectl logs -f etcd-cluster-backup-sidecar-3289547686-266b4
time="2017-09-07T23:51:18Z" level=info msg="listening on 0.0.0.0:19999"
time="2017-09-07T23:51:28Z" level=info msg="getMaxRev: member etcd-cluster-0000 revision (1)"
2017-09-07 23:51:28.924115 I | saving backup for cluster (etcd-cluster)
time="2017-09-07T23:51:29Z" level=info msg="saved backup 3.1.8_0000000000000001_etcd.backup (size: 24608) successfully"
time="2017-09-07T23:51:39Z" level=info msg="getMaxRev: member etcd-cluster-0000 revision (1)"
time="2017-09-07T23:51:39Z" level=info msg="skipped creating new backup: no change since last time"
...
```

Finally, we can check the blob listing under the Azure storage account:
```
 $ az storage blob list -c vdice-etcd-backups -o table
Name                                                        Blob Type      Length  Content Type              Last Modified
----------------------------------------------------------  -----------  --------  ------------------------  -------------------------
v1/default/etcd-cluster/3.1.8_0000000000000001_etcd.backup  BlockBlob       24608  application/octet-stream  2017-09-07T23:51:29+00:00
```

fanminshi commented:
@vdice the manual test is great. My only concern is that I don't think  `ABSContext` is needed at all. It is not being filled by anything. The current pr just uses it as an empty struct.

vdice commented:
@fanminshi OK, I'm all in favor of eliminating unused code!  Will build/deploy/test with it removed first thing tomorrow AM and report back.  Thanks!

fanminshi commented:
@vdice that will be great!

vdice commented:
@fanminshi good call; thanks for noticing that we didn't need the absContext any longer.  I've removed it and re-tested; all looks good. 👍

fanminshi commented:
lgtm

hongchaodeng commented:
@vdice  Can you add a line to CHANGELOG? LGTM after that.

vdice commented:
@hongchaodeng line to CHANGELOG added.

hongchaodeng commented:
@etcd-bot retest this please

The failure was due to known DNS issue when scaling from 1->2:

```
etcdserver: could not get cluster response from http://test-etcd-w7s0f-0000.test-etcd-w7s0f.e2e-etcd-operator-2264.svc:2380: Get http://test-etcd-w7s0f-0000.test-etcd-w7s0f.e2e-etcd-operator-2264.svc:2380/members: dial tcp: i/o timeout
```
